src/dafny/BinaryOperations.dfy

```diff
@@ -0,0 +1,202 @@
+    && IsLeftUnital(bop, unit)
+    && IsRightUnital(bop, unit)
+  }
+
+  ghost predicate IsLeftInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    forall x | inverse.requires(x) :: bop(inverse(x), x) == unit
+  }
+
+  ghost predicate IsRightInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    forall x | inverse.requires(x) :: bop(x, inverse(x)) == unit
+  }
+
+  ghost predicate IsInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    && IsLeftInverse(bop, inverse, unit)
+    && IsRightInverse(bop, inverse, unit)
+  }
+
+  ghost predicate IsAbelian<T(!new)>(bop: (T, T) -> T) {
+    forall x, y :: bop(x, y) == bop(y, x)
+  }
+
+  ghost predicate IsRightDistributive<T(!new)>(bop1: (T, T) -> T, bop2: (T, T) -> T) {
+    forall x, y, z :: bop2(bop1(x, y), z) == bop1(bop2(x, z), bop2(y, z))
+  }
+
+  ghost predicate IsLeftDistributive<T(!new)>(bop1: (T, T) -> T, bop2: (T, T) -> T) {
+    forall x, y, z :: bop2(x, bop1(y, z)) == bop1(bop2(x, y), bop2(x, z))
+  }
+
+  ghost predicate IsDistributive<T(!new)>(bop1: (T, T) -> T, bop2: (T, T) -> T) {
+    && IsLeftDistributive(bop1, bop2)
+    && IsRightDistributive(bop1, bop2)
+  }
+
+  ghost predicate IsMonoid<T(!new)>(bop: (T, T) -> T, unit: T) {
+    && IsAssociative(bop)
+    && IsUnital(bop, unit)
+  }
+
+  datatype MonoidStructure<!T(!new)> = MonoidStructure(bop: (T, T) -> T, unit: T)
+  type Monoid<!T(!new)> = g: MonoidStructure | IsMonoid(g.bop, g.unit) witness *
+
+  ghost predicate IsGroup<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    && IsMonoid(bop, unit)
+    && IsInverse(bop, inverse, unit)
+  }
+
+  datatype GroupStructure<!T(!new)> = GroupStructure(bop: (T, T) -> T, inverse: T -> T, unit: T)
+  type Group<!T(!new)> = g: GroupStructure | IsGroup(g.bop, g.inverse, g.unit) witness *
+
+  ghost predicate IsAbelianGroup<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    && IsGroup(bop, inverse, unit)
+    && IsAbelian(bop)
+  }
+
+  type AbelianGroup<!T(!new)> = g: GroupStructure | IsAbelianGroup(g.bop, g.inverse, g.unit) witness *
+
+  ghost predicate IsRing<T(!new)>(add: (T, T) -> T, minus: T -> T, zero: T, mult: (T, T) -> T, one: T) {
+    && IsAbelianGroup(add, minus, zero)
+    && IsMonoid(mult, one)
+    && IsDistributive(add, mult)
+  }
+
+  datatype RingStructure<!T(!new)> = RingStructure(add: (T, T) -> T, minus: T -> T, zero: T, mult: (T, T) -> T, one: T)
+  type Ring<!T(!new)> = r: RingStructure | IsRing(r.add, r.minus, r.zero, r.mult, r.one) witness *
+
+  ghost predicate IsField<T(!new)>(add: (T, T) -> T, minus: T -> T, zero: T, mult: (T, T) -> T, div: T --> T, one: T)
+    requires forall t :: (t != zero) ==> div.requires(t)
+  {
+    && IsAbelianGroup(add, minus, zero)
+    && IsAbelianGroup(mult, div, one)
+    && IsLeftDistributive(add, mult)
+  }
+
+  datatype FieldStructure<!T(!new)> = RingStructure(add: (T, T) -> T, minus: T -> T, zero: T, mult: (T, T) -> T, div: T --> T, one: T)
```

Comment:
Should the constructor be called FieldStructure? Or did you purposely want to reuse RingStructure?

Reply:
Good catch, I renamed it.